src/main/java/org/dcsa/tnt/service/impl/TNTEventSubscriptionTOServiceImpl.java

```diff
@@ -0,0 +1,357 @@
+                  .then(createShipmentEventType(eventSubscriptionTO))
+                  .then(createTransportEventType(eventSubscriptionTO))
+                  .then(createEquipmentEventType(eventSubscriptionTO))
+                  .thenReturn(eventSubscriptionTO);
+            });
+  }
+
+  // ToDo : replace this with mapstruct
+  private final Function<TNTEventSubscriptionTO, EventSubscription>
+      eventSubscriptionTOToEventSubscription =
+          esTo -> {
+            EventSubscription eventSubscription = new EventSubscription();
+            eventSubscription.setSubscriptionID(esTo.getSubscriptionID());
+            eventSubscription.setCallbackUrl(esTo.getCallbackUrl());
+            eventSubscription.setSecret(esTo.getSecret());
+            eventSubscription.setCarrierBookingReference(esTo.getCarrierBookingReference());
+            eventSubscription.setEquipmentReference(esTo.getEquipmentReference());
+            eventSubscription.setCarrierServiceCode(esTo.getCarrierServiceCode());
+            eventSubscription.setCarrierVoyageNumber(esTo.getCarrierVoyageNumber());
+            eventSubscription.setVesselIMONumber(esTo.getVesselIMONumber());
+            eventSubscription.setTransportDocumentReference(esTo.getTransportDocumentReference());
+            eventSubscription.setTransportCallID(esTo.getTransportCallID());
+            return eventSubscription;
+          };
+
+  @Override
+  public Mono<TNTEventSubscriptionTO> update(TNTEventSubscriptionTO eventSubscriptionTO) {
+    return eventSubscriptionRepository
+        .deleteEventTypesForSubscription(eventSubscriptionTO.getSubscriptionID())
+        .thenReturn(eventSubscriptionTO)
+        .map(eventSubscriptionTOToEventSubscription)
+        .flatMap(
+            updated ->
+                eventSubscriptionService
+                    .findById(updated.getSubscriptionID())
+                    .map(
+                        original -> {
+                          updated.setSecret(original.getSecret());
+                          updated.copyInternalFieldsFrom(original);
+                          return updated;
+                        }))
+        .flatMap(eventSubscriptionService::update)
+        .flatMap(ignored -> createEventTypes(eventSubscriptionTO));
+  }
+
+  private Mono<TNTEventSubscriptionTO> createEventTypes(
+      TNTEventSubscriptionTO eventSubscriptionTO) {
+
+    List<EventType> eventTypes;
+
+    if (CollectionUtils.isEmpty(eventSubscriptionTO.getEventType())) {
+      eventTypes = ALL_ALLOWED_EVENT_TYPES;
+      eventSubscriptionTO.setEventType(eventTypes);
+    } else {
+      eventTypes = eventSubscriptionTO.getEventType();
+    }
+
+    return Flux.fromIterable(eventTypes)
+        .concatMap(
+            eventType ->
+                eventSubscriptionRepository.insertEventTypeForSubscription(
+                    eventSubscriptionTO.getSubscriptionID(), eventType))
+        .then(Mono.just(eventSubscriptionTO));
+  }
```

Comment:
To me this should be located in EventCore with an abstract method called getAllAllowedEventTypes() which needs to be implemented in this class